### PR TITLE
Module 3

### PR DESCRIPTION
**Pull request by em4go**
Me he enfocado en crear un modelo de regresión logística y comparar las posibles variables que puede tener el modelo

### CODE REVIEW

File: src/module_3/functions.py [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+
+import pickle
+
+
+def train_val_test_split(
+    X: pd.DataFrame,
+    y: pd.Series,
+    test_size: float = 0.3,
+    random_state: int or None = None,
+) -> tuple:
```
Comment (sergiorozada12) [inline src/module_3/functions.py:22]: Tuple of what? Also, don't use native python ```tuple```, use ```Tuple``` from typing

File: src/module_3/functions.py [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+    X_train, X_test, y_train, y_test = train_test_split(
+        X, y, test_size=test_size, random_state=random_state
+    )
+    X_train, X_val, y_train, y_val = train_test_split(
```
Comment (sergiorozada12) [inline src/module_3/functions.py:26]: You are not actually controlling the size of the test size explicitly here. You have a test dataset of size ```test_size**2```.

File: src/module_3/functions.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+def filter_orders_by_n_products(df: pd.DataFrame, n_products: int) -> pd.DataFrame:
+    products_ordered = df[df["outcome"] == 1]
+    orders_len = products_ordered.groupby("order_id")["variant_id"].count()
+    orders_over_n_products = orders_len[orders_len >= n_products].index
+    filtered = df[df["order_id"].isin(orders_over_n_products)]
+    return filtered
+
+
+def add_useful_variables(df: pd.DataFrame) -> pd.DataFrame:
+    products_sorted = df.sort_values(
```
Comment (sergiorozada12) [inline src/module_3/functions.py:108]: I leave you the comment here, but applies to the NBs as well. All the variables here are pandas dataframes: ```last_ordered_products```, ```times_ordered```, etc. It is much more clear to the reader if they have the ```df``` prefix.

File: src/module_3/functions.py [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    user_id: int,
+    variant_id: int,
+    df: pd.DataFrame,
+) -> pd.DataFrame:
+    X = df.query(f"user_id == {user_id} and variant_id == {variant_id}")
+    return X
+
+
+def get_possible_models() -> list[list[str]]:
+    from possible_models import (
```
Comment (sergiorozada12) [inline src/module_3/functions.py:164]: This is a little bit clumpsy. Why not declaring a list of models in possible models? Then, you'd save this function.

File: src/module_3/functions.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        MODEL_3,
+        MODEL_4,
+        MODEL_5,
+    )
+
+    all_models = [MODEL_1, MODEL_2, MODEL_3, MODEL_4, MODEL_5]
+    return all_models
+
+
+def get_model(model_name) -> LogisticRegression:
```
Comment (sergiorozada12) [inline src/module_3/functions.py:176]: What's ```model_name```?

File: src/module_3/possible_models.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
+MODEL_1 = [
```
Comment (sergiorozada12) [inline src/module_3/possible_models.py:1]: ```possible_models.py``` is a terrible name for a file. Consider creating a ```utils/config.py``` or something similar.

File: src/module_3/milestone_1.ipynb
Comment (sergiorozada12) [inline src/module_3/milestone_1.ipynb:1]: Things here:
* Instead of copying the dfs to send them to the sklearn split, cast them directly into np.arrays. At the very end you will be copying the data, will save some operations to sklearn, and feels more natural.
* To avoid garbage outputs in a NB cell add ; at the end of the line (model.train).
* Add df to the names of the dataframes.
* Beware of absolutely good or bad conclusions. Having a 99% of whatever metric might be good, or not. What if we consistently fail in a relevant 1% of products that are crucial for the business? In retail, for example, partners can pay for product placements and we should be recommending them properly.

File: src/module_3/milestone_2.ipynb
Comment (sergiorozada12) [inline src/module_3/milestone_2.ipynb:1]: The introduction of this NB should be more thorough. I liked it better how you faced milestone 1. This is less self-contained. 

You have a very good point with respect to saving the df with preprocessed features. We call that feature stores, if you want to take a look:

https://www.hopsworks.ai/dictionary/feature-store

**Review by sergiorozada12 — APPROVED** [review]
Good PR. Take a look at the comments and you can merge.
